[ventaair] New VentaAir binding for air humidifiers

t2000 opened the pull request:
New binding that implements support for air humidifier from Venta Air.

Closes #9922

Signed-off-by: Stefan Triller <github@stefantriller.de>

<!--
Thanks for contributing to the openHAB project!
Please describe the goal and effect of your PR here.
Pay attention to the below notes and to *the guidelines* for this repository.
Feel free to delete any comment sections in the template (starting with "<!--").
-->

<!-- TITLE -->

<!--
Please provide a PR summary in the *Title* above, according to the following schema:
- If related to one specific add-on: Mention the add-on shortname in square brackets
  e.g. "[exec]", "[netatmo]" or "[tesla]"
- If the PR is work in progress: Add "[WIP]"
- Give a short meaningful description in imperative mood
  e.g. "Add support for device XYZ" or "Fix wrongly handled exception"
  for a new add-on/binding: "Initial contribution"
Examples:
- "[homematic] Improve communication with weak signal devices"
- "[timemachine][WIP] Initial contribution"
- "Update contribution guidelines on new signing rules"
-->

<!-- DESCRIPTION -->

<!--
Please give a few sentences describing the overall goals of the pull request.
Give enough details to make the improvement and changes of the PR understandable
to both developers and tech-savy users.

Please keep the following in mind:
- What is the classification of the PR, e.g. Bugfix, Improvement, Novel Addition, ... ?
- Did you describe the PRs motivation and goal?
- Did you provide a link to any prior discussion, e.g. an issue or community forum thread?
- Did you describe new features for the end user?
- Did you describe any noteworthy changes in usage for the end user?
- Was the documentation updated accordingly, e.g. the add-on README?
- Does your contribution follow the coding guidelines:
  https://www.openhab.org/docs/developer/guidelines.html
- Did you check for any (relevant) issues from the static code analysis:
  https://www.openhab.org/docs/developer/bindings/#include-the-binding-in-the-build
- Did you sign-off your work:
  https://www.openhab.org/docs/developer/contributing.html#sign-your-work
-->

<!-- TESTING -->

<!--
Your Pull Request will automatically be built and available under the following folder:
https://openhab.jfrog.io/openhab/libs-pullrequest-local/org/openhab/

It is a good practice to add a URL to your built JAR in this Pull Request description,
so it is easier for the community to test your Add-on.
If your Pull Request contains a new binding, it will likely take some time
before it is reviewed and processed by maintainers.
That said, consider submitting your Add-on in the Eclipse IoT Marketplace
See this thread for more info:
https://community.openhab.org/t/24491

Don't forget to submit a thread about your Add-on in the openHAB community:
https://community.openhab.org/c/add-ons 
-->

Skinah commented on bundles/org.openhab.binding.ventaair/README.md:
```diff
@@ -0,0 +1,119 @@
+To do so, the binding listens to UDP port 48000 for data and creates `DiscoveryResult`s based on the received data from the device.
+This comes in handy for getting the MAC address for the device for example.
+Once the `DiscoveryResult` is added as a `Thing`, a connection to the device will be created and it will beep, showing a confirmation screen that the device "openHAB" would like to get access.
+After confirming this request, the use can link its `Item`s to receive data or control the device.
```
```suggestion
After confirming this request, the user can link items to receive data or control the device.
```

Skinah commented on bundles/org.openhab.binding.ventaair/README.md:
```diff
@@ -0,0 +1,119 @@
+## Thing Configuration
+
+There are three mandatory configuration parameters for a thing: `ipAddress`, `macAddress` and `deviceType`.
+Whereas the first two are self-exploring, the later one defines the type of the device. It is an integer value and its best to use the automatic discovery to obtain it from the device.
```
```suggestion
Whereas the first two are self-explaining, the later one defines the type of the device.
It is an integer value and its best to use the automatic discovery to obtain it from the device.
```

How many device types are there? Have you considered making multiple thing types instead?

t2000 replied:
Yes, of course I have thought about it. However, I do a) not own any other device and b) I thought a more general solution to catch ALL devices would be best for the normal user.

That's why there are just the 2 thing types. One for the device that I own and a generic one.

Skinah commented on bundles/org.openhab.binding.ventaair/README.md:
```diff
@@ -0,0 +1,119 @@
+There are three mandatory configuration parameters for a thing: `ipAddress`, `macAddress` and `deviceType`.
+Whereas the first two are self-exploring, the later one defines the type of the device. It is an integer value and its best to use the automatic discovery to obtain it from the device.
+The optional parameter `pollingTime` is the time interval in seconds in which the data should be polled from the device, default is 10 seconds.
+The `hash` parameter is also optional. It is a negative integer value and it is used by the device to identify a connection to a client, like the App from the vendor for example.
```
It may be better to have this information in a table format, check the other bindings on how they do this. You have done it for the channels already which is good. Also each sentence should be on its own line.

Skinah commented on bundles/org.openhab.binding.ventaair/README.md:
```diff
@@ -0,0 +1,119 @@
+Whereas the first two are self-exploring, the later one defines the type of the device. It is an integer value and its best to use the automatic discovery to obtain it from the device.
+The optional parameter `pollingTime` is the time interval in seconds in which the data should be polled from the device, default is 10 seconds.
+The `hash` parameter is also optional. It is a negative integer value and it is used by the device to identify a connection to a client, like the App from the vendor for example.
+I do not know whether there devices that are restricted to only one client, so I added this parameter to allow the user to set the same value as his App on the phone (can be obtained via sniffing the network).
```
```suggestion
I do not know whether their devices are restricted to only one client, so I added this parameter to allow the user to set the same value as the app on their phone (can be obtained via sniffing the network).
```

Skinah commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/VentaAirDeviceConfiguration.java:
```diff
@@ -0,0 +1,32 @@
+ */
+@NonNullByDefault
+public class VentaAirDeviceConfiguration {
+    public @Nullable String ipAddress;
+    public @Nullable String macAddress;
+    public @Nullable BigDecimal deviceType;
+    public @Nullable BigDecimal hash;
+    public @Nullable BigDecimal pollingTime;
```
```suggestion
    public String ipAddress="";
    public String macAddress="";
    public BigDecimal deviceType =BigDecimal.ZERO;
    public BigDecimal hash =BigDecimal.ZERO;
    public BigDecimal pollingTime=BigDecimal.ZERO;
```
Doing this means you don't need to check for null anymore.

Skinah commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/VentaThingHandler.java:
```diff
@@ -0,0 +1,338 @@
+        if (config.deviceType == null) {
+            return "Device Type not set, use discovery to find the correct one";
+        }
+        if (config.pollingTime == null || config.pollingTime.compareTo(BigDecimal.ZERO) < 0) {
```
```suggestion
        if (config.pollingTime == null || config.pollingTime.compareTo(BigDecimal.ZERO) <= 0) {
```
If it must be more than zero, consider what happens if it is exactly 0.

Skinah commented on bundles/org.openhab.binding.ventaair/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,32 @@
+			<label>Device Type</label>
+			<description>Type of the device as integer</description>
+		</parameter>
+		<parameter name="pollingTime" type="integer" required="false">
+			<label>Polling interval (s)</label>
```
```suggestion
		<parameter name="pollingTime" type="integer" required="true" min="1" max="6000" unit="s">
			<label>Polling Interval</label>
```
Since your using a method called validateConfig() to ensure that the value is not null and above 0, you can do it here with these changes.

Skinah commented:
I have looked at around 40% so far and it looks good so thanks for the submission.

If you fix all the above and are waiting for more feedback have a look at the 'target' folder after you compile and there is a sub folder called 'code-analysis'. Look in there at the report and fix all issues as well as any compiler warnings.

t2000 commented:
@Skinah Thank you very much for your review. I have addressed your comments with a new commit.

> have a look at the 'target' folder after you compile and there is a sub folder called 'code-analysis'. Look in there at the report and fix all issues as well as any compiler warnings

I have already fixed most of these warning, except the "@NonNullByDefault" annotations for pure data classes which only exist for the sole purpose of being filled by GSON. I do not see the point in annotating these as well as polluting the rest of the code with annotations resulting from this.

BTW: I am the guy that started to introduce the null annotations in the first place, see https://github.com/eclipse-archived/smarthome/pull/3906 ;-) So I think they sometimes help to improve the code, but in the case for pure data classes I do not see the need.

Skinah commented on bundles/org.openhab.binding.ventaair/README.md:
```diff
@@ -0,0 +1,127 @@
+
+There are three mandatory configuration parameters for a thing: `ipAddress`, `macAddress` and `deviceType`.
+
+| parameter  | mandatory/optional description |
```
```suggestion
| parameter  | mandatory/optional | description |
```

You could also use "required" as the heading and a Y and N, up to you.

Skinah commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/VentaAirDeviceConfiguration.java:
```diff
@@ -0,0 +1,32 @@
+    public BigDecimal deviceType = BigDecimal.ZERO;
+    // we all know that 42 is the answer to everything, so let's pick this one ;)
+    public BigDecimal hash = new BigDecimal("-42");
+    public BigDecimal pollingTime = new BigDecimal(10);
```
```suggestion
    public BigDecimal pollingTime = BigDecimal.TEN;
```

Skinah commented on bundles/org.openhab.binding.ventaair/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,155 @@
+	<channel-type id="discReplaceTime">
+		<item-type>Number</item-type>
+		<label>Hygiene Disc replacement</label>
+		<description>Time in hours until Hygiene Disc schould be replaced</description>
```
```suggestion
		<description>Time in hours until the Hygiene Disc should be replaced</description>
```

Skinah commented on bundles/org.openhab.binding.ventaair/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,155 @@
+
+	<channel-type id="discReplaceTime">
+		<item-type>Number</item-type>
+		<label>Hygiene Disc replacement</label>
```
```suggestion
		<label>Hygiene Disc Replacement</label>
```

Skinah commented on bundles/org.openhab.binding.ventaair/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,155 @@
+	<channel-type id="operationTime">
+		<item-type>Number</item-type>
+		<label>Operation Time</label>
+		<description>Operation Time</description>
```
This is not enough to know clearly what it does. Example:
Is this how long it will run for next time it is started?
Is this how long it has already been running for?

Skinah commented on bundles/org.openhab.binding.ventaair/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,155 @@
+
+	<channel-type id="cleaningTime">
+		<item-type>Number</item-type>
+		<label>CleaningTime</label>
```
```suggestion
		<label>Cleaning Time</label>
```

Skinah commented on bundles/org.openhab.binding.ventaair/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,155 @@
+
+	<channel-type id="waterLevel">
+		<item-type>Number</item-type>
+		<label>WaterLevel</label>
```
```suggestion
		<label>Water Level</label>
```

Skinah commented on bundles/org.openhab.binding.ventaair/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,155 @@
+
+	<channel-type id="childLock">
+		<item-type>Switch</item-type>
+		<label>ChildLock</label>
```
```suggestion
		<label>Child Lock</label>
```

Skinah commented on bundles/org.openhab.binding.ventaair/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,155 @@
+
+	<channel-type id="sleepMode">
+		<item-type>Switch</item-type>
+		<label>SleepMode</label>
```
```suggestion
		<label>Sleep Mode</label>
```

Skinah commented on bundles/org.openhab.binding.ventaair/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,155 @@
+	<channel-type id="sleepMode">
+		<item-type>Switch</item-type>
+		<label>SleepMode</label>
+		<description>SleepMode</description>
```
```suggestion
		<description>Sleep Mode</description>
```

Skinah commented on bundles/org.openhab.binding.ventaair/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,155 @@
+
+	<channel-type id="fanSpeed">
+		<item-type>Number</item-type>
+		<label>FanSpeed</label>
```
```suggestion
		<label>Fan Speed</label>
```

Skinah commented on bundles/org.openhab.binding.ventaair/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,32 @@
+			<description>Type of the device as integer</description>
+		</parameter>
+		<parameter name="pollingTime" type="integer" required="false" min="1" max="86400">
+			<label>Polling interval (s)</label>
```
```suggestion
			<label>Polling Interval (s)</label>
```
Labels should have each first letter as a capital.

Skinah commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/message/action/AutomaticAction.java:
```diff
@@ -0,0 +1,28 @@
+ */
+public class AutomaticAction implements Action {
+    @SuppressWarnings("unused")
+    private boolean Automatic;
```
variables should not start with a capital. See dev guidelines.

t2000 replied:
See below, this is because of the GSON mapping.

Skinah commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/message/action/AllActions.java:
```diff
@@ -0,0 +1,102 @@
+ *
+ */
+public class AllActions implements Action {
+    private boolean Power = true;
+    private int FanSpeed;
+    private int TargetHum = 65;
+    private int Timer;
+    private boolean Boost;
+    private boolean SleepMode;
+    private boolean ChildLock;
+    private boolean Automatic;
+    private int SysLanguage; // 3?
+    private int CleanLanguage; // 0?
+    private int TempUnit; // 0=Celsius, 1=Fahrenheit?
+    private int DisplayLeft;
+    private int DisplayRight;
+    private int Reset;
+    private int ConINet;
+    private boolean DelUser; // default false
```
All variables should use lowerCamelCase and not start with a capitial letter.

t2000 replied:
Yes, that is because they are needed for GSON. of course I could add an annotation to map all these names, but this would IMHO just clutter the code.

Skinah commented:
That is probably the limit that I can help with.
Regarding pure data classes, I am sure the next person that does a review can help on if it is ok to suppress the warning with an annotation.

t2000 commented:
Thank you for the review again. I have addressed your changes and commented on the GSON mapping.

Skinah reviewed (approved):
LGTM

fwolter commented on bundles/org.openhab.binding.ventaair/README.md:
```diff
@@ -0,0 +1,127 @@
+| channel  | type (RO=read-only)   | description                  |
+|----------|--------|------------------------------|
+| power          | Switch | This is the power on/off channel  |
+| fanSpeed       | Number | This is the channel to control the steps for the speed of the fan |
```
You could mention the range or unit of the Channel.

fwolter commented on bundles/org.openhab.binding.ventaair/README.md:
```diff
@@ -0,0 +1,127 @@
+|----------|--------|------------------------------|
+| power          | Switch | This is the power on/off channel  |
+| fanSpeed       | Number | This is the channel to control the steps for the speed of the fan |
+| targetHumidity | Number | This channel sets the target humidity that should be try to reach by the device  |
```
You could mention the range and allowed values.

fwolter commented on bundles/org.openhab.binding.ventaair/README.md:
```diff
@@ -0,0 +1,127 @@
+| power          | Switch | This is the power on/off channel  |
+| fanSpeed       | Number | This is the channel to control the steps for the speed of the fan |
+| targetHumidity | Number | This channel sets the target humidity that should be try to reach by the device  |
+| timer          | Number | This channel sets the power off timer to the set value in hours, i.e. 3 = turn off in 3 hours from now |
```
You could mention the range and allowed values.

fwolter commented on bundles/org.openhab.binding.ventaair/README.md:
```diff
@@ -0,0 +1,127 @@
+| automatic      | Switch | This is the control channel to start the automatic operation mode of the device |
+| cleanMode      | Switch (RO) | This is the channel that indicates if the device is in the cleaning mode |
+| temperature    | Number:Temperature (RO) | This channel provides the current measured temperature in Celsius or Fahrenheit as configured on the device |
+| humidity       | Number (RO) | This channel provides the humidity measured by the device in percent |
```
You could use `QuantityType` here with `Units.PERCENT`.
```suggestion
| humidity       | Number:Dimenionsless (RO) | This channel provides the humidity measured by the device in percent |
```

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/Communicator.java:
```diff
@@ -0,0 +1,179 @@
+            OutputStream output = socket.getOutputStream();
+
+            byte[] dataToSend = buildMessageBytes(messageJson, "GET", "Complete");
+            logger.debug("Sending request data message (String):\n{}", new String(dataToSend));
+            logger.debug("Sending request data message (bytes): [{}]", HexUtils.bytesToHex(dataToSend, ", "));
```
Can this message be replaced by using the debugger? See point 4 https://www.openhab.org/docs/developer/guidelines.html#f-logging

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/Communicator.java:
```diff
@@ -0,0 +1,179 @@
+            logger.debug("Sending request data message (bytes): [{}]", HexUtils.bytesToHex(dataToSend, ", "));
+            output.write(dataToSend);
+
+            logger.debug("Data was sent to outputstream");
```
See above. Debugger.

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/Communicator.java:
```diff
@@ -0,0 +1,179 @@
+
+            logger.debug("Data was sent to outputstream");
+
+            BufferedReader br = new BufferedReader(new InputStreamReader(input));
```
The `BufferedReader` should be closed.

Can you specify the encoding you expect? E.g. StandardCharsets.UTF_8

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/Communicator.java:
```diff
@@ -0,0 +1,179 @@
+                if (reply.startsWith("{")) {
+                    // remove padding byte(s) after JSON data
+                    String data = String.valueOf(reply.toCharArray(), 0, reply.length() - 1);
+                    logger.debug("Got Data: {}", data);
```
See above. Debugger. Please check all.

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/Communicator.java:
```diff
@@ -0,0 +1,179 @@
+            }
+            socket.close();
+        } catch (IOException e) {
+            logger.debug("Communication problem while polling data from device", e);
```
Are you logging the stack trace by intention?

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/Communicator.java:
```diff
@@ -0,0 +1,179 @@
+
+    private byte[] buildMessageBytes(String message, String method, String endpoint) throws IOException {
+        ByteArrayOutputStream getInfoOutputStream = new ByteArrayOutputStream();
+        getInfoOutputStream.write(createMessageHeader(method, endpoint, message.length()).getBytes());
```
See above. Specify encoding.

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/Communicator.java:
```diff
@@ -0,0 +1,179 @@
+    public void stopPollDataFromDevice() {
+        ScheduledFuture<?> localPollingJob = pollingJob;
+        if (localPollingJob != null && !localPollingJob.isCancelled()) {
+            logger.debug("Canceling polling job");
```
See above. Debugger. Same for below.

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/VentaThingHandler.java:
```diff
@@ -0,0 +1,335 @@
+
+    private final Logger logger = LoggerFactory.getLogger(VentaThingHandler.class);
+
+    @NonNullByDefault({})
+    private VentaAirDeviceConfiguration config;
```
You could initialize the field with an empty configuration object to get rid of the null disabling.

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/VentaThingHandler.java:
```diff
@@ -0,0 +1,335 @@
+            case VentaAirBindingConstants.CHANNEL_POWER:
+                if (command instanceof OnOffType) {
+                    OnOffType power = (OnOffType) command;
+                    dispatchActionToDevice(new PowerAction(power == OnOffType.ON ? true : false));
```
If you do a comparison only, you don't need to cast the command and neither check the type.

Same for below.
```suggestion
                    dispatchActionToDevice(new PowerAction(command == OnOffType.ON));
```

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/VentaThingHandler.java:
```diff
@@ -0,0 +1,335 @@
+            try {
+                localCommunicator.sendActionToDevice(action);
+            } catch (IOException e) {
+                updateStatus(ThingStatus.OFFLINE, ThingStatusDetail.COMMUNICATION_ERROR);
```
You could include the exception's message.

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/VentaThingHandler.java:
```diff
@@ -0,0 +1,335 @@
+            Unit<Temperature> temperatureUnit = SIUnits.CELSIUS;
+
+            if (actions != null) {
+                OnOffType powerState = actions.isPower() ? OnOffType.ON : OnOffType.OFF;
```
```suggestion
                OnOffType powerState = OnOffType.from(actions.isPower());
```

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/discovery/VentaDeviceDiscovery.java:
```diff
@@ -0,0 +1,176 @@
+
+    @Override
+    protected void startScan() {
+        logger.debug("startScan() pressed");
```
See above. Debugger. Please check all.

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/discovery/VentaDeviceDiscovery.java:
```diff
@@ -0,0 +1,176 @@
+        logger.debug("startBackgroundDiscovery()");
+
+        ScheduledFuture<?> localScanJob = scanJob;
+        if (localScanJob != null && !localScanJob.isCancelled()) {
```
No harm in canceling a task that is already cancelled.
```suggestion
        if (localScanJob != null) {
```

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/discovery/VentaDeviceDiscovery.java:
```diff
@@ -0,0 +1,176 @@
+            createDiscoveryResult(thingTypeUID, m.getHeader());
+
+        } catch (IOException e) {
+            logger.debug("Exception while opening port or trying to parse discovery data from device.", e);
```
See above. Stacktrace.

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/discovery/VentaDeviceDiscovery.java:
```diff
@@ -0,0 +1,176 @@
+        Gson gson = new Gson();
+        Message msg = null;
+
+        String packetAsString = new String(packet);
```
See above. Encoding.

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/discovery/VentaDeviceDiscovery.java:
```diff
@@ -0,0 +1,176 @@
+
+        String typeLabel = thingTypeUID.getId().toUpperCase();
+
+        DiscoveryResult result = DiscoveryResultBuilder.create(uid).withRepresentationProperty("macAddress")
```
You could make `macAddress` a constant.

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/discovery/VentaDeviceDiscovery.java:
```diff
@@ -0,0 +1,176 @@
+        String typeLabel = thingTypeUID.getId().toUpperCase();
+
+        DiscoveryResult result = DiscoveryResultBuilder.create(uid).withRepresentationProperty("macAddress")
+                .withProperties(properties).withLabel(typeLabel + " (IP=" + ipAddress + " Mac=" + macAddress + ")")
```
The label should be as short as possible. Is the MAC address necessary?

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/message/CommandMessage.java:
```diff
@@ -0,0 +1,33 @@
+ * @author Stefan Triller - Initial contribution
+ *
+ */
+public class CommandMessage extends Message {
```
You could move DTOs in a package called `dto` to get rid of the checkstyle warnings about missing null annotations.

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,32 @@
+			<description>Type of the device as integer</description>
+		</parameter>
+		<parameter name="pollingTime" type="integer" required="false" min="1" max="86400">
+			<label>Polling Interval (s)</label>
```
```suggestion
			<label>Polling Interval</label>
```

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,32 @@
+			<label>Device Type</label>
+			<description>Type of the device as integer</description>
+		</parameter>
+		<parameter name="pollingTime" type="integer" required="false" min="1" max="86400">
```
```suggestion
		<parameter name="pollingTime" type="integer" required="false" unit="s" min="1" max="86400">
```

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/resources/OH-INF/config/config.xml:
```diff
@@ -0,0 +1,32 @@
+			<default>10</default>
+			<description>Time in seconds between fetching data from the device</description>
+		</parameter>
+		<parameter name="hash" type="integer" required="false">
```
```suggestion
		<parameter name="hash" type="integer" required="false" max="-1">
```

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/resources/OH-INF/thing/channels.xml:
```diff
@@ -0,0 +1,155 @@
+		<item-type>Number</item-type>
+		<label>Fan RPM</label>
+		<description>Fan RPM</description>
+		<state pattern="%d" readOnly="true"/>
```
```suggestion
		<state pattern="%d RPM" readOnly="true"/>
```

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/resources/OH-INF/thing/thing-types.xml:
```diff
@@ -0,0 +1,67 @@
+
+	<!-- Generic type has boost channel and maybe in the future other channels -->
+	<thing-type id="generic">
+		<label>Generic VentaAir Humidifier/Cleaner</label>
```
Labels are expected to be as short as possible. Guideline is 2-3 words with max 20-25 chars. See https://www.openhab.org/docs/developer/bindings/thing-xml.html#formatting-labels-and-descriptions
```suggestion
		<label>Generic Humidifier/Cleaner</label>
```

fwolter commented on bundles/org.openhab.binding.ventaair/README.md:
```diff
@@ -0,0 +1,127 @@
+| humidity       | Number (RO) | This channel provides the humidity measured by the device in percent |
+| waterLevel     | Number (RO) | This channel indicates the water level of the tank where 1 is equal to the yellow "refill tank" warning on the device/App |
+| fanRPM         | Number (RO) | This channel provides the speed of the ventilation fan |
+| timerTimePassed  | Number (RO) | If a timer has been set, this channel provides the minutes since when the timer was started  |
```
Same for below.
```suggestion
| timerTimePassed  | Number:Time (RO) | If a timer has been set, this channel provides the minutes since when the timer was started  |
```

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/Communicator.java:
```diff
@@ -0,0 +1,179 @@
+ */
+@NonNullByDefault
+public class Communicator {
+    private final Duration COMMUNICATION_TIMEOUT = Duration.ofSeconds(5);
```
```suggestion
    private static final Duration COMMUNICATION_TIMEOUT = Duration.ofSeconds(5);
```

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/discovery/VentaDeviceDiscovery.java:
```diff
@@ -0,0 +1,176 @@
+
+    private final Logger logger = LoggerFactory.getLogger(VentaDeviceDiscovery.class);
+
+    private final Duration TIME_BETWEEN_SCANS = Duration.ofSeconds(30);
```
Please move all static fields to the top.
```suggestion
    private static final Duration TIME_BETWEEN_SCANS = Duration.ofSeconds(30);
```

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/discovery/VentaDeviceDiscovery.java:
```diff
@@ -0,0 +1,176 @@
+    private void findDevices() {
+        byte[] buf = new byte[512];
+        try (DatagramSocket socket = new DatagramSocket(VentaAirBindingConstants.PORT)) {
+
```
See checkstyle warning. Remove empty line after brace.

fwolter commented on bundles/org.openhab.binding.ventaair/src/main/java/org/openhab/binding/ventaair/internal/discovery/VentaDeviceDiscovery.java:
```diff
@@ -0,0 +1,176 @@
+                    break;
+            }
+            createDiscoveryResult(thingTypeUID, m.getHeader());
+
```
Remove empty line before brace.

fwolter reviewed (changes requested):
@Skinah already did a good job! Here are my review comments.

t2000 commented:
Thank you very much for the second review.

I have addressed your changes in a new commit and rebased my branch against the latest main branch.

fwolter commented on bundles/org.openhab.binding.ventaair/README.md:
```diff
@@ -0,0 +1,127 @@
+| automatic      | Switch | This is the control channel to start the automatic operation mode of the device |
+| cleanMode      | Switch (RO) | This is the channel that indicates if the device is in the cleaning mode |
+| temperature    | Number:Temperature (RO) | This channel provides the current measured temperature in Celsius or Fahrenheit as configured on the device |
+| humidity       | Number:Dimenionsless (RO) | This channel provides the humidity measured by the device in percent |
```
```suggestion
| humidity       | Number:Dimensionless (RO) | This channel provides the humidity measured by the device in percent |
```

t2000 commented:
I have fixed the typo using github but now the dco check is complaining. Can you ignore that or do I have to use git magic to change the commit?

fwolter reviewed (approved):
LGTM

I forced the DCO to pass for the GitHub based changed.

t2000 commented:
Since this PR has 2 approvals and no errors/failed checks, is there anything that is missing from my side until this gets merged? If so, please let me know. Thank you.

kaikreuzer commented:
With @fwolter's approval, we should be good to merge indeed. Thanks for your contribution!
